cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/spec/StageSpec.java

```diff
@@ -53,6 +56,7 @@ private StageSpec(String name, PluginSpec plugin, Map<String, Schema> inputSchem
     this.name = name;
     this.plugin = plugin;
     this.inputSchemas = Collections.unmodifiableMap(inputSchemas);
+    this.inputStages = Lists.newArrayList(inputSchemas.keySet());
```

Comment:
This seems a bit unsafe to me. The original problem was that the schemas map keyset doesn't contains all input stages. And now we are using the same keyset to construct the input stages list?

I think it is better to have the stage names pass in explicitly (probably with a fallback to the schemas keyset for backward compatibility when deserializing from older spec).

Reply:
fixed.